### PR TITLE
Add some API for Xen command-line

### PR DESCRIPTION
**Pull request by wjn740**
add_grub_xen_cmdline_settings
replace_grub_xen_cmdline_settings
remove_grub_xen_cmdline_settings

Xen kernel's command-line will make from "GRUB_CMDLINE_XEN_DEFAULT".

### CODE REVIEW

**Comment by wjn740** [comment]
@pevik Hi Petr, please help review.

**Comment by wjn740** [comment]
> Thanks for notifying me. How about using `get_cmdline_var()` as a default parameter instead of copy and paste?
> 
> ```
> sub add_grub_cmdline_settings {
>     my ($add, $update_grub, $search) = @_;
>     $search //= get_cmdline_var();
>     change_grub_config('"$', " $add\"", $search, "g", $update_grub);
> }
> sub add_grub_xen_cmdline_settings {
>     my ($add, $update_grub) = @_;
>     add_grub_cmdline_settings($add, $update_grub, "GRUB_CMDLINE_XEN_DEFAULT");
> }
> 
> sub replace_grub_cmdline_settings {
>     my ($old, $new, $update_grub, $search) = @_;
>     $search //= get_cmdline_var();
>     change_grub_config($old, $new, $search, "g", $update_grub);
> }
> sub replace_grub_xen_cmdline_settings {
>     my ($old, $new, $update_grub) = @_;
>     replace_grub_cmdline_settings($old, $new, $update_grub, "GRUB_CMDLINE_XEN_DEFAULT");
> }
> ```

Hi sir,
For now, get_cmdline_var() only support linux kernel. It could be used get line of linux parameter. not xen.
Because when we running in xen Dom0, we might need to modify both linux command-line and xen command-line.

**Comment by wjn740** [comment]
Sorry, I misunderstand your words, let me post the next version to see. @pevik

**Comment by wjn740** [comment]
Hi @pevik, I think this version fine to review.

**Comment by pevik** [comment]
@wjn740 my way would be this (suppose that passing just `$remove` to `remove_grub_cmdline_settings()` will lead `$search` to be undef and thus used default value in `replace_grub_cmdline_settings()`):
```
sub remove_grub_cmdline_settings {
    my ($remove, $search) = @_;
    replace_grub_cmdline_settings('[[:blank:]]*' . $remove . '[[:blank:]]*', " ", "g", $search);
}
sub remove_grub_xen_cmdline_settings {
    my $remove = shift;
    remove_grub_cmdline_settings($remove, "GRUB_CMDLINE_XEN_DEFAULT");
}
```
I don't like copy&paste strange parameters `'[[:blank:]]*' . $remove . '[[:blank:]]*', " ", "g"`, better looks to me reuse functions and just copy&paste `GRUB_CMDLINE_XEN_DEFAULT`. But maybe I'm wrong, maybe was your first version really more readable.

**Comment by wjn740** [comment]
OK. Let me try to the thrid version.

**Comment by wjn740** [comment]
Hi, @pevik Please review this version.

File: lib/bootloader_setup.pm [hunk 1]
```diff
@@ -1040,37 +1040,77 @@ sub change_grub_config {
...
 
 Add C<$add> into /etc/default/grub, using sed.
 C<$update_grub> if set, regenerate /boot/grub2/grub.cfg with grub2-mkconfig and upload configuration.
+C<$search> if set, bypass default grub cmdline variable.
 =cut
 sub add_grub_cmdline_settings {
+    my ($add, $update_grub, $search) = @_;
+    $search //= get_cmdline_var();
+    change_grub_config('"$', " $add\"", $search, "g", $update_grub);
+}
```
Comment (pevik) [inline lib/bootloader_setup.pm]: I didn't notice, that using `$search` will require to specify `$update_grub`, maybe it'd make sense to use `%args` for these. But I leave it up to you.

Comment (wjn740) [reply lib/bootloader_setup.pm]: I think I need to fix this issue so that it could be used safely. Please waiting the next version.

Comment (pevik) [reply lib/bootloader_setup.pm]: Thanks for taking care! And sorry to bother you so much for just simple functions.

Comment (wjn740) [reply lib/bootloader_setup.pm]: No, I'm just perl junior, I have to learn something. and Product usually crashes at the simple functions.

File: lib/bootloader_setup.pm [hunk 2]
```diff
@@ -1040,37 +1040,95 @@ sub change_grub_config {
...
 C<$update_grub> if set, regenerate /boot/grub2/grub.cfg with grub2-mkconfig and upload configuration.
+C<$search> if set, bypass default grub cmdline variable.
 =cut
 sub add_grub_cmdline_settings {
+    my (%args) = compat_args(
+        {
+            add         => '',
+            update_grub => 0,
+            search      => get_cmdline_var();
+        }, ['add', 'update_grub', 'search'], @_);
```
Comment (pevik) [inline lib/bootloader_setup.pm]: Good point to use `compat_args` :).

File: lib/bootloader_setup.pm [hunk 3]
```diff
@@ -1040,37 +1040,95 @@ sub change_grub_config {
...
+
+    my $add         = %args{add};
+    my $update_grub = %args{update_grub};
+    my $search      = %args{search};
```
Comment (pevik) [inline lib/bootloader_setup.pm]: You need to use `$` (scalar vs list context): `$args{priority}`

File: lib/bootloader_setup.pm [hunk 4]
```diff
@@ -1040,37 +1040,95 @@ sub change_grub_config {
...
+    change_grub_config('"$', " $add\"", $search, "g", $update_grub);
```
Comment (pevik) [inline lib/bootloader_setup.pm]: I'd use `$args` directly:
`change_grub_config('"$', " $args{add}\"", $args{search}, "g", $args{update_grub})`

File: lib/bootloader_setup.pm (hunk 3 above)
Comment (pevik) [reply lib/bootloader_setup.pm]: Also travis told you that :)
https://travis-ci.org/os-autoinst/os-autoinst-distri-opensuse/jobs/581481735
`Global symbol "%args" requires explicit package name at lib/bootloader_setup.pm line 1057.`

It's good to 1) do run checks locally with `make tidy` 2) do code verification (safes time for reviewer).

**Comment by wjn740** [comment]
Yes, I have noticed by CI. I am across working on another task.  so I will fix build error later. :-)

File: lib/bootloader_setup.pm (hunk 3 above)
Comment (wjn740) [reply lib/bootloader_setup.pm]: sorry, I mark a WIP for avoiding time waste. When It is ready, I will ping you.

**Comment by pevik** [comment]
ok :)

**Comment by wjn740** [comment]
Hi @pevik, how about this version?

File: lib/bootloader_setup.pm [hunk 5]
```diff
@@ -1040,37 +1040,90 @@ sub change_grub_config {
...
+Search line C<$search> from /etc/default/grub (use for sed).
 =cut
 sub remove_grub_cmdline_settings {
+    my ($remove, $search) = @_;
+    replace_grub_cmdline_settings('[[:blank:]]*' . $remove . '[[:blank:]]*', " ", "g", $search);
+}
+
+=head2 remove_grub_xen_cmdline_settings
+
+    remove_grub_xen_cmdline_settings($remove);
```
Comment (pevik) [inline lib/bootloader_setup.pm]: `remove_grub_xen_cmdline_settings($remove, [ $search ]);`
\+ maybe note, that when `$search` is undef, it's used `replace_grub_cmdline_settings()` default (that's probably obvious).

Comment (wjn740) [reply lib/bootloader_setup.pm]: Let me fix them.

File: lib/bootloader_setup.pm [hunk 6]
```diff
@@ -1040,37 +1040,91 @@ sub change_grub_config {
...
 
-    remove_grub_cmdline_settings($remove);
+    remove_grub_cmdline_settings($remove [, $search]);
 
 Remove C<$remove> from /etc/default/grub (using sed) and regenerate /boot/grub2/grub.cfg.
+Search line C<$search> from /etc/default/grub (use for sed).
 =cut
 sub remove_grub_cmdline_settings {
+    my ($remove, $search) = @_;
+    $search //= get_cmdline_var();
```
Comment (pevik) [inline lib/bootloader_setup.pm]: I wouldn't add `$search //= get_cmdline_var();` here. It's not wrong, but it's a duplicity - it can be handled only in `replace_grub_cmdline_settings()`. WDYT?

**Comment by wjn740** [comment]
let me try again.

File: lib/bootloader_setup.pm (hunk 6 above)
Comment (wjn740) [reply lib/bootloader_setup.pm]: because you mention when search is 'undef', so perl can handle 'undef' parameter passing?

**Comment by wjn740** [comment]
@pevik , this version I remove the wrong get_cmdline_var().  add $search in grep_grub_cmdline_setting.

File: lib/bootloader_setup.pm [hunk 7]
```diff
@@ -1003,14 +1003,15 @@ sub grep_grub_settings {
...
-    grep_grub_cmdline_settings($pattern)
+    grep_grub_cmdline_settings($pattern [, $search])
 
 Search for C<$pattern> in grub cmdline variable (usually
 GRUB_CMDLINE_LINUX_DEFAULT) in /etc/default/grub, return 1 if found.
 =cut
 sub grep_grub_cmdline_settings {
     my $pattern = shift;
-    return grep_grub_settings(get_cmdline_var() . ".*${pattern}");
+    my $search //= get_cmdline_var();
```
Comment (pevik) [inline lib/bootloader_setup.pm]: James, you miss to shift for `$search` (so `get_cmdline_var()` will be used all the time). That's the only concern.
I guess there will be always just 2 params, so we `%args` wouldn't make much sense here.

Comment (wjn740) [reply lib/bootloader_setup.pm]: Yes, 3+ parameters use %args make sense only. 

let me try again.

**Comment by wjn740** [comment]
@pevik , please review this version.

**Comment by wjn740** [comment]
> Thanks for your work and patience.

You are a good teacher.
Happy teacher's day. :-)

https://en.wikipedia.org/wiki/List_of_Teachers%27_Days

**Comment by pevik** [comment]
> You are a good teacher.
> Happy teacher's day. :-)

:).

@asmorodskyi @jalausuch can you please merge it?

File: lib/bootloader_setup.pm (hunk 5 above)
Comment (asmorodskyi) [reply lib/bootloader_setup.pm]: looks like you haven't push the fix here

Comment (pevik) [reply lib/bootloader_setup.pm]: @asmorodskyi I consider this PR ready, just nobody merged it (I don't have merge rights for openQA).

File: lib/bootloader_setup.pm (hunk 2 above)
Comment (mloviska) [reply lib/bootloader_setup.pm]: but compat_args aren't exported by testapi.


It broke JeOS completely :)
https://openqa.suse.de/tests/3451320#step/grub2_gfxmode/10

**Comment by mloviska** [comment]
guys, where are the VRs ? I cannot see them :)

**Comment by asmorodskyi** [comment]
@mloviska  - sorry for that , I will revert this one after travis check - #8620

**Comment by mloviska** [comment]
> @mloviska - sorry for that , I will revert this one after travis check - #8620

no problem, thanks Anton :)

**Comment by asmorodskyi** [comment]
@wjn740  - your PR got reverted because it introduced regression 
https://openqa.suse.de/tests/3451320#step/grub2_gfxmode/10 . please open new PR , fix issue mentioned by @mloviska and do VRs :) and I will merge it again

**Comment by pevik** [comment]
I'm sorry to ack broken code. @asmorodskyi thanks for handling this.

**Comment by wjn740** [comment]
So sorry about broken progress. I will review it. to see how to do that.

**Comment by wjn740** [comment]
@pevik about this issue, what do you suggest? Export a function? or do a manual parameter check?

**Comment by mloviska** [comment]
> @pevik about this issue, what do you suggest? Export a function? or do a manual parameter check?

you can always just put the package name in front
```perl 
testapi::compat_args
```
